config/initializers/devise.rb

```diff
@@ -326,3 +326,10 @@
   # so you need to do it manually. For the users scope, it would be:
   # config.omniauth_path_prefix = '/my_engine/users/auth'
 end
+
+Warden::Manager.after_set_user do |user, auth, opts|
+  if user.roles.empty?
+    auth.logout
+    throw(:warden)
+  end
+end
```

Comment:
This little change deals with the situation where a user has no roles (in that case we don't let them log in). This mirrors the behavior of dealing with discarded users (that is done with a default scope on users `default_scope { kept }`). 

I considered 2 other options:
- adding to the default scope to filter out users without roles but I was worried this could affect existing behavior
- attempted to modify the `create` method of the session controller but that resulted in a confusing error for the user (it would seem like they logged in and got logged out immediately). I needed to address the logic in the `super` method.